publisher-sdk/src/main/java/com/criteo/publisher/model/DeviceInfo.java

```diff
@@ -17,51 +17,43 @@
 package com.criteo.publisher.model;
 
 import android.content.Context;
-import android.text.TextUtils;
-import android.webkit.WebView;
+import android.os.Build.VERSION;
+import android.os.Build.VERSION_CODES;
+import android.webkit.WebSettings;
 import androidx.annotation.NonNull;
 import androidx.annotation.UiThread;
 import androidx.annotation.VisibleForTesting;
 import com.criteo.publisher.SafeRunnable;
-import com.criteo.publisher.concurrent.RunOnUiThreadExecutor;
-import com.criteo.publisher.logging.Logger;
-import com.criteo.publisher.logging.LoggerFactory;
 import com.criteo.publisher.util.CompletableFuture;
 import com.criteo.publisher.util.PreconditionsUtil;
+import java.util.concurrent.Executor;
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 public class DeviceInfo {
 
-  @NonNull
-  private final Logger logger = LoggerFactory.getLogger(getClass());
-
   @NonNull
   private final Context context;
 
   @NonNull
-  private final RunOnUiThreadExecutor runOnUiThreadExecutor;
+  private final Executor executor;
 
   @NonNull
   private final CompletableFuture<String> userAgentFuture = new CompletableFuture<>();
 
   @NonNull
   private final AtomicBoolean isInitialized = new AtomicBoolean(false);
 
-  public DeviceInfo(@NonNull Context context, @NonNull RunOnUiThreadExecutor runOnUiThreadExecutor) {
+  public DeviceInfo(@NonNull Context context, @NonNull Executor executor) {
     this.context = context;
-    this.runOnUiThreadExecutor = runOnUiThreadExecutor;
+    this.executor = executor;
   }
 
   public void initialize() {
-    // This needs to be run on UI thread because a WebView is used to fetch the user-agent
-    runOnUiThread(new Runnable() {
-      @Override
-      public void run() {
-        if (isInitialized.compareAndSet(false, true)) {
-          String userAgent = resolveUserAgent();
-          userAgentFuture.complete(userAgent);
-        }
+    runSafely(() -> {
+      if (isInitialized.compareAndSet(false, true)) {
```

Comment:
could you add a condition just before the runSafely to return earlier if the flag isInitialized is already true.
It is fine to have 2 threads concurrently trying to challenge for the isInitialized flag. But it is quite useless to post new async task if the user agent is already init.